src/metabase/mbql/util.clj

```diff
@@ -0,0 +1,139 @@
+(ns metabase.mbql.util
+  "Utilitiy functions for working with MBQL queries."
+  (:require [clojure
+             [string :as str]
+             [walk :as walk]]
+            [metabase.mbql.schema :as mbql.s]
+            [metabase.util :as u]
+            [metabase.util.schema :as su]
+            [schema.core :as s]))
+
+(s/defn normalize-token :- s/Keyword
+  "Convert a string or keyword in various cases (`lisp-case`, `snake_case`, or `SCREAMING_SNAKE_CASE`) to a lisp-cased
+  keyword."
+  [token :- su/KeywordOrString]
+  (-> (u/keyword->qualified-name token)
+      str/lower-case
+      (str/replace #"_" "-")
+      keyword))
+
+(defn mbql-clause?
+  "True if `x` is an MBQL clause (a sequence with a keyword as its first arg). (Since this is used by the code in
+  `normalize` this handles pre-normalized clauses as well.)"
+  [x]
+  (and (sequential? x)
+       (keyword? (first x))))
+
+(defn is-clause?
+  "If `x` an MBQL clause, and an instance of clauses defined by keyword(s) `k-or-ks`?
+
+    (is-clause? :count [:count 10])        ; -> true
+    (is-clause? #{:+ :- :* :/} [:+ 10 20]) ; -> true"
+  [k-or-ks x]
+  (and
+   (mbql-clause? x)
+   (if (coll? k-or-ks)
+     ((set k-or-ks) (first x))
+     (= k-or-ks (first x)))))
+
+(defn clause-instances
+  "Return a sequence of all the instances of clause(s) in `x`. Like `is-clause?`, you can either look for instances of a
+  single clause by passing a single keyword or for instances of multiple clauses by passing a set of keywords. Returns
+  `nil` if no instances were found.
+
+    ;; look for :field-id clauses
+    (clause-instances :field-id {:query {:filter [:= [:field-id 10] 20]}})
+    ;;-> [[:field-id 10]]
+
+    ;; look for :+ or :- clauses
+    (clause-instances #{:+ :-} ...)"
+  {:style/indent 1}
+  [k-or-ks x]
+  (let [instances (transient [])]
```

Comment:
I think you want an atom here, from the [transients reference](https://clojure.org/reference/transients)
```
Note in particular that transients are not designed to be bashed in-place. You must capture and use the return value in the next call.
```

Reply:
TIL. I will fix